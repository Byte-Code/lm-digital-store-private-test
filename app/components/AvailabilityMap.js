import React, { Component, PropTypes } from 'react';
import GoogleMapReact from 'google-map-react';
import { meters2ScreenPixels } from 'google-map-react/utils';
import ImmutablePropTypes from 'react-immutable-proptypes';
import styled from 'styled-components';
import PlaceIcon from 'material-ui/svg-icons/maps/place';
import MyPlaceIcon from 'material-ui/svg-icons/maps/person-pin-circle';
import BlockIcon from 'material-ui/svg-icons/navigation/close';
import Slider from 'material-ui/Slider';

import { getStockLabel } from '../utils/utils';

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  color: #fff;
`;

const Map = styled.div`
  width: 100%;
  height: 1080px;
  position: relative;
`;

const Title = styled.h1`
  font-size: 48px;
  text-align: center;
  line-height: 70px;
`;

const ProductInfo = styled.h3`
  font-size: 14px;
  text-align: center;
  text-transform: capitalize;
  margin-bottom: 66px;
`;

export const StoreBadge = styled.div`
  display: flex;
  padding: 10px;
  margin-bottom: 20px;
  cursor: pointer;
  background-color: ${props => (props.isActive ? 'rgba(255, 255, 255, 0.2)' : 'transparent')};
`;

const InfoWindow = styled.div`
  width: 300px;
  background-color: #fff;
  box-shadow: 0 0 8px 0 #e4e4e4;
  position: absolute;
  top: -195px;
  left: 0;
  padding: 20px;
  z-index: 1;
`;

const Divider = styled.div`
  border: 1px dashed #333333;
  width: 100%;
`;

const Stock = styled.div`
  margin-top: 15px;
  color: ${props => (props.isAvailable ? '#67cb33' : '#333333')}
  font-size: 14px;
  font-family: LeroyMerlinSans Bold;
`;

const StoreName = styled.p`
  font-size: 20px;
  color: #262626;
  margin-bottom: 10px;
  font-family: LeroyMerlinSans Bold;
`;

const ArrowDown = styled.div`
  width: 0;
  height: 0;
  border-left: 14px solid transparent;
  border-right: 14px solid transparent;
  border-top: 14px solid #fff;
  position: absolute;
  bottom: -14px;
  left: 14px;
  filter: drop-shadow(0 2px 0 #e4e4e4);
`;

const Address = styled.div`
  margin-bottom: 20px;
  &>p {
    font-size: 15px;
    color: #333333;
  }
`;

const WhiteBg = styled.div`
  position: absolute;
  height: 25px;
  width: 20px;
  z-index: -1;
  top: 9px;
  left: 16px;
  background-color: #fff;
`;

const Icon = styled.div`
  position: relative;
  z-index: 1;
`;

const Circle = styled.div`
  box-sizing: content-box;
  border: 5px dotted #67cb33;
  transform: translate(-50%, -50%);
  border-radius: 50%;
  box-shadow: 0 0 0 9999px rgba(0, 0, 0, 0.3);
  pointer-events: none;
}
`;

export default class AvailabilityMap extends Component {
  static propTypes = {
    productName: PropTypes.string.isRequired,
    productCode: PropTypes.string.isRequired,
    nearbyStoreStock: ImmutablePropTypes.list.isRequired,
    currentStore: ImmutablePropTypes.map.isRequired
  }

  constructor(props) {
    super(props);
    this.state = {
      selectedStore: null,
      infoWindowOpen: false,
      zoom: 11
    };
  }

  selectStore = (storeCode) => {
    this.setState({ selectedStore: storeCode, infoWindowOpen: true });
  }

  closeInfoWindow = () => {
    this.setState({ infoWindowOpen: false });
  }

  updateZoom = (e) => {
    this.setState({ zoom: e.zoom });
  }

  renderMarkers() {
    const { nearbyStoreStock, currentStore } = this.props;
    return nearbyStoreStock
    .map(s => {
      const currentStock = s.get('storeStock');
      const code = s.get('code');
      const lat = s.get('latitude');
      const lng = s.get('longitude');
      const isCurrentStore = currentStore.get('code') === code;
      const MarkerIcon = isCurrentStore ? MyPlaceIcon : PlaceIcon;
      const iconColor = currentStock > 0 ? '#67cb33' : '#000';

      return (
        <Icon
          key={code}
          lat={lat}
          lng={lng}
        >
          <MarkerIcon
            style={{ height: 55, width: 55, cursor: 'pointer' }}
            color={iconColor}
            onTouchTap={() => this.selectStore(code)}
          />
          <WhiteBg />
        </Icon>
      );
    });
  }

  // TODO move this logic into a separate component
  renderInfoWindow() {
    const { nearbyStoreStock } = this.props;
    const { selectedStore, infoWindowOpen } = this.state;
    if (!infoWindowOpen || !selectedStore) {
      return null;
    }
    const currentStoreStockInfo = nearbyStoreStock.find(ns => ns.get('code') === selectedStore);
    const name = currentStoreStockInfo.get('name');
    const availability = currentStoreStockInfo.get('storeStock');
    const isAvailable = availability > 0;
    const label = getStockLabel(availability);
    const street = currentStoreStockInfo.getIn(['address', 'street']);
    const streetNumber = currentStoreStockInfo.getIn(['address', 'streetNumber']) || '';
    const zip = currentStoreStockInfo.getIn(['address', 'zipCode']);
    const city = currentStoreStockInfo.getIn(['address', 'city']);
    const province = currentStoreStockInfo.getIn(['address', 'state']);
    const lat = currentStoreStockInfo.get('latitude');
    const lng = currentStoreStockInfo.get('longitude');

    return (
      <InfoWindow lat={lat} lng={lng}>
        <BlockIcon
          style={{
            width: 24,
            height: 24,
            position: 'absolute',
            top: 5,
            right: 5,
            cursor: 'pointer'
          }}
          fill="#333333"
          onTouchTap={this.closeInfoWindow}
        />
        <div>
          <StoreName>{name}</StoreName>
          <Address>
            <p>{`${street} ${streetNumber}`}</p>
            <p>{`${city} - (${province})`}</p>
            <p>{zip}</p>
          </Address>
        </div>
        <Divider />
        <Stock isAvailable={isAvailable}>
          {label}
        </Stock>
        <ArrowDown />
      </InfoWindow>
    );
  }

  render() {
    const { productName, productCode, currentStore } = this.props;
    const { zoom } = this.state;
    const lat = currentStore.getIn(['gpsInformation', 'x']);
    const lng = currentStore.getIn(['gpsInformation', 'y']);
    const center = { lat, lng };
    const { w, h } = meters2ScreenPixels(80000, { lat, lng }, zoom);

    return (
      <Wrapper>
        <Title>Verifica Disponibilit√†</Title>
        <ProductInfo>{`${productName} - REF. ${productCode}`}</ProductInfo>
        <Map>
          <GoogleMapReact
            center={center}
            defaultZoom={11}
            fullscreenControl={false}
            options={{ fullscreenControl: false }}
            onChange={this.updateZoom}
          >
            {this.renderMarkers()}
            {this.renderInfoWindow()}
            <Circle lat={lat} lng={lng} style={{ width: parseInt(h, 10), height: parseInt(w, 10) }} />
          </GoogleMapReact>
        </Map>
      </Wrapper>
    );
  }
}
