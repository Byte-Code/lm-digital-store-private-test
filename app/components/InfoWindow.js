import React from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import BlockIcon from 'material-ui/svg-icons/navigation/close';
import glamorous from 'glamorous';

import { getStockLabel } from '../utils/utils';

const Wrapper = glamorous.div({
  width: '300px',
  backgroundColor: '#fff',
  boxShadow: '0 0 8px 0 #e4e4e4',
  position: 'absolute',
  top: '-230px',
  left: 0,
  padding: '20px',
  zIndex: 1
});

const Divider = glamorous.div({
  border: '1px dashed #333333',
  width: '100%'
});

const Stock = glamorous.div(({ isAvailable }) => ({
  marginTop: '15px',
  color: isAvailable ? '#67cb33' : '#333333',
  fontSize: '14px',
  fontFamily: 'LeroyMerlinSans Bold'
}));

const StoreName = glamorous.p({
  fontSize: '20px',
  color: '#262626',
  marginBottom: '10px',
  fontFamily: 'LeroyMerlinSans Bold'
});

const ArrowDown = glamorous.div({
  width: 0,
  height: 0,
  borderLeft: '14px solid transparent',
  borderRight: '14px solid transparent',
  borderTop: '14px solid #fff',
  position: 'absolute',
  bottom: '-14px',
  left: '14px',
  filter: 'drop-shadow(0 2px 0 #e4e4e4)'
});

const Address = glamorous.div({
  marginBottom: '20px',
  '&>p': {
    fontSize: '15px',
    color: '#333333'
  }
});

const iconStyle = {
  width: 24,
  height: 24,
  position: 'absolute',
  top: 5,
  right: 5,
  cursor: 'pointer'
};

const InfoWindow = ({ currentStoreInfo, handleClick }) => {
  const name = currentStoreInfo.get('name');
  const availability = currentStoreInfo.get('storeStock');
  const isAvailable = availability > 0;
  const label = getStockLabel(availability);
  const street = currentStoreInfo.getIn(['address', 'street']);
  const streetNumber = currentStoreInfo.getIn(['address', 'streetNumber']) || '';
  const zip = currentStoreInfo.getIn(['address', 'zipCode']);
  const city = currentStoreInfo.getIn(['address', 'city']);
  const province = currentStoreInfo.getIn(['address', 'state']);

  return (
    <Wrapper>
      <BlockIcon style={iconStyle} fill="#333333" onTouchTap={handleClick} />
      <div>
        <StoreName>{name}</StoreName>
        <Address>
          <p>{`${street} ${streetNumber}`}</p>
          <p>{`${city} - (${province})`}</p>
          <p>{zip}</p>
        </Address>
      </div>
      <Divider />
      <Stock isAvailable={isAvailable}>
        {label}
      </Stock>
      <ArrowDown />
    </Wrapper>
  );
};

InfoWindow.propTypes = {
  currentStoreInfo: ImmutablePropTypes.map.isRequired,
  handleClick: PropTypes.func.isRequired
};

export default InfoWindow;
