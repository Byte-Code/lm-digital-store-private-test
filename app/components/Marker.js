import React, { PropTypes } from 'react';
import PlaceIcon from 'material-ui/svg-icons/maps/place';
import MyPlaceIcon from 'material-ui/svg-icons/maps/person-pin-circle';
import glamorous from 'glamorous';

const Wrapper = glamorous.div({
  position: 'relative',
  zIndex: 1,
  transform: 'translate(-50%, -50%)'
});

const WhiteBg = glamorous.div({
  position: 'absolute',
  height: '25px',
  width: '20px',
  zIndex: -1,
  top: '9px',
  left: '16px',
  backgroundColor: '#fff'
});


const Marker = ({ handleClick, isCurrentStore, isAvailable }) => {
  const Icon = isCurrentStore ? MyPlaceIcon : PlaceIcon;
  const iconColor = isAvailable ? '#67cb33' : '#000';
  return (
    <Wrapper>
      <Icon
        style={{ height: 55, width: 55, cursor: 'pointer' }}
        color={iconColor}
        onTouchTap={handleClick}
      />
      <WhiteBg />
    </Wrapper>
  );
};

Marker.propTypes = {
  isCurrentStore: PropTypes.bool.isRequired,
  isAvailable: PropTypes.bool.isRequired,
  handleClick: PropTypes.func.isRequired
};

export default Marker;
