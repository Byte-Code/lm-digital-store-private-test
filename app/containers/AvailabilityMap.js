import { connect } from 'react-redux';

import AvailabilityMap from '../components/AvailabilityMap';
import { getNearbyStores, getStore } from '../reducers/selectors';

const mapStateToProps = (state, ownProps) => {
  const { allStoreStock } = ownProps;
  const nearbyStores = getNearbyStores(state);
  const nearbyStoreStock = nearbyStores.map(s => {
    const hasProduct = allStoreStock.find(ns => ns.get('storeCode') === s.get('code'));
    if (hasProduct) {
      return s.set('storeStock', hasProduct.get('storeStock'));
    } return s.set('storeStock', 0);
  });
  const currentStore = getStore(state);
  return { nearbyStoreStock, currentStore };
};

export default connect(mapStateToProps, {})(AvailabilityMap);
