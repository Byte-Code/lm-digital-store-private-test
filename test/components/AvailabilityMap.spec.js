import React from 'react';
import { fromJS } from 'immutable';
import { shallow } from 'enzyme';

import mountWithStore from '../../app/utils/testingUtils';
import AvailabilityMap from '../../app/components/AvailabilityMap';
import Marker from '../../app/components/Marker';
import InfoWindow from '../../app/components/InfoWindow';

// TODO find a way to test user interactions
const productCode = '123';
const productName = 'abc';
const nearbyStoreStock = fromJS([
  {
    code: '123',
    name: 'abc',
    latitude: 1231,
    longitude: 123213,
    address: {
      state: 'abc',
      street: 'bde',
      streetNumber: '1',
      zipCode: '213',
      city: 'milan'
    }
  },
  {
    code: '1234',
    name: 'asdsa',
    latitude: 12311,
    longitude: 123213,
    address: {
      state: 'asd',
      street: 'bae',
      streetNumber: '1',
      zipCode: '213',
      city: 'roma'
    }
  }
]);
const currentStore = fromJS({
  code: '123',
  name: 'current',
  gpsInformation: {
    x: 12311,
    y: 123213
  }
});
const requestFetchNearbyStores = jest.fn();

describe('AvailabilityDialog', () => {
  const shallowWrapper = shallow(
    <AvailabilityMap
      productCode={productCode}
      productName={productName}
      nearbyStoreStock={nearbyStoreStock}
      currentStore={currentStore}
      requestFetchNearbyStores={requestFetchNearbyStores}
    />
  );

  it('should initialize with the right interal state', () => {
    const expectedState = {
      selectedStore: null,
      infoWindowOpen: false,
      zoom: 11,
      radius: 30
    };
    expect(shallowWrapper.state()).toEqual(expectedState);
    expect(shallowWrapper).toMatchSnapshot();
  });

  it('should render a marker for each nearbyStore in radius', () => {
    expect(shallowWrapper.find(Marker)).toHaveLength(2);
  });

  it('should not have rendered InfoWindow', () => {
    expect(shallowWrapper.find(InfoWindow)).toHaveLength(0);
  });

  it('should call requestFetchNearbyStores with the right args after mounting', () => {
    mountWithStore(
      <AvailabilityMap
        productCode={productCode}
        productName={productName}
        nearbyStoreStock={nearbyStoreStock}
        currentStore={currentStore}
        requestFetchNearbyStores={requestFetchNearbyStores}
      />
    );
    expect(requestFetchNearbyStores).toHaveBeenCalledWith(12311,
      123213, 30);
  });
});
