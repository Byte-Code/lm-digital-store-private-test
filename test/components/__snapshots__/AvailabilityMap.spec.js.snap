// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvailabilityDialog should initialize with the right interal state 1`] = `
<glamorous(div)>
  <glamorous(h1)>
    Verifica Disponibilit√†
  </glamorous(h1)>
  <glamorous(h3)>
    abc - REF. 123
  </glamorous(h3)>
  <glamorous(p)>
    Seleziona il raggio di distanza dal negozio di current
  </glamorous(p)>
  <Slider
    axis="x"
    color="#67cb33"
    disableFocusRipple={false}
    disabled={false}
    max={50}
    min={10}
    onChange={[Function]}
    required={true}
    sliderStyle={
      Object {
        "marginBottom": 24,
      }
    }
    step={10}
    style={Object {}}
    value={30}
  />
  <glamorous(div)>
    30 km
  </glamorous(div)>
  <glamorous(div)>
    <GoogleMap
      center={
        Object {
          "lat": 12311,
          "lng": 123213,
        }
      }
      debounced={true}
      defaultZoom={11}
      distanceToMouse={[Function]}
      fullscreenControl={false}
      googleMapLoader={[Function]}
      hoverDistance={30}
      layerTypes={Array []}
      onChange={[Function]}
      options={
        Object {
          "fullscreenControl": false,
          "maxZoom": 17,
          "minZoom": 8,
          "zoomControl": false,
        }
      }
      style={
        Object {
          "height": "100%",
          "margin": 0,
          "padding": 0,
          "position": "relative",
          "width": "100%",
        }
      }
      yesIWantToUseGoogleMapApiInternals={false}
    >
      <Component
        code="123"
        handleClick={[Function]}
        isAvailable={false}
        isCurrentStore={true}
        lat={1231}
        lng={123213}
      />
      <Component
        code="1234"
        handleClick={[Function]}
        isAvailable={false}
        isCurrentStore={false}
        lat={12311}
        lng={123213}
      />
      <glamorous(div)
        height={2405.6967429351644}
        lat={12311}
        lng={123213}
        width={2403.821601331234}
      />
    </GoogleMap>
  </glamorous(div)>
</glamorous(div)>
`;
